activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -520,10 +520,7 @@ def find_some_ordered(ids)
         result = except(:limit, :offset).where(primary_key => ids).records
 
         if result.size == ids.size
-          pk_type = @klass.type_for_attribute(primary_key)
-
-          records_by_id = result.index_by(&:id)
-          ids.map { |id| records_by_id.fetch(pk_type.cast(id)) }
+          result.in_order_of(:id, ids.map { |id| @klass.type_for_attribute(primary_key).cast(id) })
         else
```

Comment:
Just a guest's guess, but shouldn't `@klass.type_for_attribute(primary_key)` be taken out of the `map` block, since this is 4 more unnecessary method calls per iteration?

Reply:
Also and maybe as a suggestion for a future improvements: it certainly would be great to apply that ordering in the database already.

Reply:
Didn't consider the micro optimization worth it given that the type_for_attribute call is a simple type that exits early.